columnflow/production/cms/dy.py

```diff
@@ -0,0 +1,189 @@
+# coding: utf-8
+
+"""
+Column production methods related to Drell-Yan reweighting.
+"""
+
+from __future__ import annotations
+import functools
+
+import law
+
+from dataclasses import dataclass
+
+from columnflow.production import Producer, producer
+from columnflow.util import maybe_import, InsertableDict
+from columnflow.columnar_util import set_ak_column
+
+np = maybe_import("numpy")
+ak = maybe_import("awkward")
+vector = maybe_import("vector")
+
+
+# helper
+set_ak_column_f32 = functools.partial(set_ak_column, value_type=np.float32)
+
+
+logger = law.logger.get_logger(__name__)
+
+
+@dataclass
+class DrellYanConfig:
+    era: str
+    correction: str
+    unc_correction: str
+
+    def __post_init__(self) -> None:
+        if not self.era or not self.correction or not self.unc_correction:
+            raise ValueError("Campaign era, correction  and unc_correction must be set")
+
+
+@producer(
+    uses={"GenPart.*"},
+    produces={"gen_dilepton.pt"},
+)
+def get_gen_dilepton(self, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Reconstruct the di-lepton pair from generator level info.
+    """
+
+    # get the absolute pdg id (to account for anti-particles) and status of the particles
+    pdg_id = abs(events.GenPart.pdgId)
+    status = events.GenPart.status
+
+    # electrons and muons need to have status == 1, so find them
+    ele_mu_mask = (pdg_id == 11) | (pdg_id == 13)
+    final_ele_mu_mask = ele_mu_mask & (status == 1)
+
+    # taus need to have status == 2,
+    tau_mask = pdg_id == 15
+    final_tau_mask = tau_mask & (status == 2)
+
+    # combine the masks
+    lepton_mask = final_ele_mu_mask | final_tau_mask
+    # only consider leptons from hard process (i.e. from the matrix element)
+    mask_gen = ak.mask(events.GenPart, lepton_mask).hasFlags("fromHardProcess")
+
+    # fill the mask with False if it is None and extract the gen leptons
+    mask_gen = ak.fill_none(mask_gen, False)
+    lepton_pairs = events.GenPart[mask_gen]
+
+    # some up the four momenta of the leptons
+    lepton_pair_momenta = lepton_pairs.sum(axis=-1)
+
+    # finally, save the pt of the lepton pair on generator level
+    events = set_ak_column(events, "gen_dilepton.pt", lepton_pair_momenta.pt)
+    return events
+
+
+@producer(
+    uses={get_gen_dilepton.PRODUCES},
```

Comment:
Since the gen lepton producer is running during the selection, and the weight producer most likely during the ProduceColumns stage later, there shouldn't be a direct dependence, but we can just add the requested column name here.

```suggestion
    uses={"gen_dilepton_pt"},
```